File: leetcode/1052-grumpy-bookstore-owner/maxSatisfied.go
package leetcode

func maxSatisfied(customers []int, grumpy []int, minutes int) int {
	var result = 0

	for index, customer := range customers {
		if grumpy[index] == 0 {
			result += customer
		}
	}

	var (
		increase    = 0
		maxIncrease = 0
	)

	for i := 0; i < minutes; i++ {
		increase = increase + customers[i]*grumpy[i]
	}

	maxIncrease = increase

	for i := minutes; i < len(customers); i++ {
		increase =
			increase -
				customers[i-minutes]*grumpy[i-minutes] +
				customers[i]*grumpy[i]

		maxIncrease = max(maxIncrease, increase)
	}

	return result + maxIncrease
}

File: leetcode/1052-grumpy-bookstore-owner/maxSatisfied.rs
use crate::Solution;

impl Solution {
    pub fn max_satisfied(customers: Vec<i32>, grumpy: Vec<i32>, minutes: i32) -> i32 {
        let mut result = 0;

        for (index, customer) in customers.iter().enumerate() {
            if grumpy[index] == 0 {
                result = result + customer;
            }
        }

        let mut increase = 0;
        let mut max_increase = 0;

        for i in 0..minutes as usize {
            increase = increase + customers[i] * grumpy[i]
        }
        max_increase = increase;

        for i in minutes..customers.len() as i32 {
            increase = increase
                - customers[(i - minutes) as usize] * grumpy[(i - minutes) as usize]
                + customers[i as usize] * grumpy[i as usize];

            max_increase = max_increase.max(increase);
        }

        return result + max_increase;
    }
}
